doc/developer/design/20220516_transactional_consistency.md

```diff
@@ -0,0 +1,416 @@
+
+##### Do Nothing
+
+The “Do Nothing” approach involves doing nothing to solve this problem. We just allow reads to block if the global
+timestamp is too high.
+
+**Pros:**
+
+- Simple to implement.
+- Does not limit write throughput.
+
+**Cons:**
+
+- Allows unbounded delay on reads.
+
+##### Block Writes
+
+Before performing a write that would advance the global timestamp ahead of the current wall time, we block the write
+until the current wall time catches up to the global timestamp. As an optimization we can implement a form of group
+commit, where all writes within the same millisecond commit together at the same timestamp.
+
+**Pros:**
+
+- Prevents reads from being blocked.
+
+**Cons:**
+
+- Limits write throughput (1 write per millisecond without group commit) or limits write latency (1 millisecond per
+  write with group commit).
+
+##### Hybrid Logical Timestamps
+
+Note: There might be an actual name for this concept, but I chose the name above because it’s similar to Hybrid Logical
+Clocks.
+
+Currently, the timestamps used to in Materialize for `upper`, `since`, and the global timestamp are 64 bit unsigned
+integers that are interpreted as an epoch in milliseconds. Instead we can change the representation of these timestamps
+so that the X most significant bits are used for the epoch and the Y least significant bits are used for a counter.
+Every 1 millisecond the X bits are increased by 1 and the Y bytes are all reset to 0. Every write increments the Y bits
+by 1. This would allow us to have 2^Y writes per millisecond, without blocking any queries. Any write in
+excess of 2^Y per millisecond must be blocked until the next millisecond.
+
+Note: Increasing the size of the timestamp type would give more flexibility (for example to 128 bits).
+
+**Pros:**
+
+- With a sufficiently high Y, no queries would be blocked.
+
+**Cons:**
+
+- Complicated to implement and likely has a large blast radius.
+- If X is low enough we will overflow our timestamps at some point in the future. This can be mitigated with a high
+  enough X.
+
+##### Write Ahead Log (WAL)
+
+This approach involves the coordinator buffering all writes in a durable write ahead log. All writes that occur in the
+same millisecond can be assigned the same timestamp, as long as there are no reads in-between them. Then at the end of
+the millisecond, or after X milliseconds, all writes are sent to STORAGE in a batch. If Materialize crashes/restarts
+before sending the writes to STORAGE, then it can just read the unsent write from the durable logs after starting back
+up. This prevents us from having to increment the global timestamp on every write. However, this does not solve the
+problem for read/write cycles.
+
+**Pros:**
+
+- No queries are blocked from consecutive writes.
+- A WAL would be useful for many other reasons not in scope for this document. It’s likely that no matter what solution
+  we pick, we’d still want to implement a WAL.
+
+**Cons:**
```

Comment:
Requires the WAL backend to have a distributed consensus operation.

Reply:
Added. Is there somewhere I can read to see why this is required?

Reply:
https://github.com/MaterializeInc/materialize/pull/11345/files is my personal summary of this with the various options we considered. It rounds down to: Kyle brought up table transaction questions then helped us think through some possible solutions, and using a WAL correctly was where we landed. I *think* it's this way because the WAL was assumed to be highly available (distributed), and thus must have a consensus operation. But it turns out distributed consensus is just very hard, so using a WAL puts some strict limits on what we can use. However, this is mostly storage/persist's problem, because we'd most probably use persist as the WAL.